pkg/compactor/partition_compaction_planner.go

```diff
@@ -2,30 +2,146 @@ package compactor
+	visitMarkerManager := NewVisitMarkerManager(p.bkt, p.logger, p.ringLifecyclerID, visitMarker)
+	existingPartitionVisitMarker := &partitionVisitMarker{}
+	err := visitMarkerManager.ReadVisitMarker(p.ctx, existingPartitionVisitMarker)
+	visitMarkerExists := true
+	if err != nil {
+		if errors.Is(err, errorVisitMarkerNotFound) {
+			visitMarkerExists = false
+		} else {
+			return nil, fmt.Errorf("unable to get visit marker file for partition with partition ID %d, partitioned group ID %d: %s", partitionID, partitionedGroupID, err.Error())
+		}
+	}
+	if visitMarkerExists {
+		if existingPartitionVisitMarker.GetStatus() == Completed {
+			p.compactorMetrics.compactionsNotPlanned.WithLabelValues(p.userID, cortexMetaExtensions.TimeRangeStr()).Inc()
+			level.Warn(p.logger).Log("msg", "partition is in completed status", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "compactor_id", p.ringLifecyclerID, existingPartitionVisitMarker.String())
+			return nil, plannerCompletedPartitionError
+		}
+		if !existingPartitionVisitMarker.IsPendingByCompactor(p.partitionVisitMarkerTimeout, partitionID, p.ringLifecyclerID) {
+			p.compactorMetrics.compactionsNotPlanned.WithLabelValues(p.userID, cortexMetaExtensions.TimeRangeStr()).Inc()
+			level.Warn(p.logger).Log("msg", "partition is not visited by current compactor", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "compactor_id", p.ringLifecyclerID, existingPartitionVisitMarker.String())
+			return nil, plannerVisitedPartitionError
+		}
+	}
+
+	// Ensure all blocks fits within the largest range. This is a double check
+	// to ensure there's no bug in the previous blocks grouping, given this Plan()
+	// is just a pass-through.
+	// Modified from https://github.com/cortexproject/cortex/pull/2616/files#diff-e3051fc530c48bb276ba958dd8fadc684e546bd7964e6bc75cef9a86ef8df344R28-R63
+	largestRange := p.ranges[len(p.ranges)-1]
+	rangeStart := getRangeStart(metasByMinTime[0], largestRange)
+	rangeEnd := rangeStart + largestRange
+	noCompactMarked := p.noCompBlocksFunc()
+	resultMetas := make([]*metadata.Meta, 0, len(metasByMinTime))
+
+	for _, b := range metasByMinTime {
+		if b.ULID == DUMMY_BLOCK_ID {
+			continue
+		}
+		blockID := b.ULID.String()
+		if _, excluded := noCompactMarked[b.ULID]; excluded {
+			continue
+		}
+
+		if b.MinTime < rangeStart || b.MaxTime > rangeEnd {
+			return nil, fmt.Errorf("block %s with time range %d:%d is outside the largest expected range %d:%d", blockID, b.MinTime, b.MaxTime, rangeStart, rangeEnd)
+		}
+
+		resultMetas = append(resultMetas, b)
+	}
+
+	if len(resultMetas) < 1 {
+		level.Info(p.logger).Log("msg", "result meta size is empty", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "size", len(resultMetas))
+		return nil, nil
+	}
+
+	go visitMarkerManager.HeartBeat(p.ctx, errChan, p.partitionVisitMarkerFileUpdateInterval, false)
```

Comment:
Why we dont want to delete on exit? Is it to give a chance for the same compactor to pick up compaction?
Is it dangerous to delete on exit and just let all race for the plan?

Reply:
We should keep visit marker. The status information is used by https://github.com/cortexproject/cortex/blob/master/pkg/compactor/partitioned_group_info.go#L115. Also, cleaner would use those information to find out which parent blocks can be deleted.

Reply:
Ah makes sense. That is how the cleaner also knows what to do. OK